src/util/DeferredProperty.java

```diff
@@ -0,0 +1,231 @@
+     */
+    private final Supplier<T> supplier;
+
+    /**
+     * The collection of {@link DeferredProperty} instances that should be made dirty when this instance gets dirty.
+     */
+    private final List<DeferredProperty<?>> observers;
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     * If no initial value is provided (i.e. <code>null</code> value) then this instance will start of as dirty.
+     * Otherwise, if a none <code>null</code> value is provided then it will start of clean.
+     *
+     * @param value The initial value of this property.
+     * @param supplier The non-parametric factory method for updating this property value automatically.
+     * @param observers The dependent observers to which marking this instance as dirty is propagated to.
+     */
+    public DeferredProperty(T value, Supplier<T> supplier, List<DeferredProperty<?>> observers) {
+        this.value = value;
+        this.supplier = supplier;
+        this.observers = observers;
+        isDirty = value == null;
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param supplier The non-parametric factory method for updating this property value automatically.
+     */
+    public DeferredProperty(T value, Supplier<T> supplier) {
+        this(value, supplier, new ArrayList<>());
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param observers The dependent observers to which marking this instance as dirty is propagated to.
+     */
+    public DeferredProperty(T value, List<DeferredProperty<?>> observers) {
+        this(value, null, observers);
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     * @param observers The dependent observers to which marking this instance as dirty is propagated to.
+     */
+    public DeferredProperty(T value, DeferredProperty<?> ... observers) {
+        this(value, List.of(observers));
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>}.
+     *
+     * @param value The initial value of this property.
+     */
+    public DeferredProperty(T value) {
+        this(value, new ArrayList<>());
+    }
+
+    /**
+     * Constructs an instance of a {@link DeferredProperty<T>} with an initial value of <code>null</code> and no observers.
+     */
+    public DeferredProperty() {
+        this(null, new ArrayList<>());
+    }
+
+    /**
+     * Bypasses the dirty checks and just returns the value of {@link DeferredProperty#value}.
+     *
+     * @return The current value of {@link DeferredProperty#value}.
+     */
+    public T getValue() {
```

Comment:
```suggestion
    public T getRawValue() {
```

I guess in normal use people should call `get()` for getting the value that takes `isDirty()` into account. But the function name `get()` does not specify what it is getting. So I suggest to change `get()` to `getValue()` and `getValue()` to `getRawValue()` (or maybe `getUncheckedValue()`).